go/vt/vtgate/tabletgateway.go

```diff
@@ -105,31 +105,65 @@ func NewTabletGateway(ctx context.Context, hc discovery.HealthCheck, serv srvtop
 		localCell:         localCell,
 		retryCount:        *RetryCount,
 		statusAggregators: make(map[string]*TabletStatusAggregator),
-		buffer:            buffer.New(),
 	}
-	// subscribe to healthcheck updates so that buffer can be notified if needed
-	// we run this in a separate goroutine so that normal processing doesn't need to block
-	hcChan := hc.Subscribe()
-	bufferCtx, bufferCancel := context.WithCancel(ctx)
-	go func(ctx context.Context, c chan *discovery.TabletHealth, buffer *buffer.Buffer) {
-		for {
-			select {
-			case <-ctx.Done():
-				return
-			case result := <-hcChan:
-				if result == nil {
-					// If result is nil it must mean the channel has been closed. Stop goroutine in that case
-					bufferCancel()
+	gw.setupBuffering(ctx)
+	gw.QueryService = queryservice.Wrap(nil, gw.withRetry)
```

Comment:
I see we rely on srv keyspace to detect shard split events, how would a failover being handled here? Say we have a failover in a shard from A to B:
t1: A became problematic / not responding
t2: Orchestrator detects the problem and do an external reparent from A to B
t3: healthcheck detect the reparent event and set primary of the shard to B

After t3, requests should be handled properly. I'm thinking before t3, what is the mechanism to handle / buffer requests (if it exists)? I'm asking because if we can buffer as much requests as possible between t1 and t3, in theory we should have higher availability. 

Is there some coordination we are doing at vttablet to trigger the buffer logic here for a reparent?